File: answers/lintcode/Longest Consecutive Sequence.cpp
class Solution {
public:
    /**
     * @param nums: A list of integers
     * @return an integer
     */
    int longestConsecutive(vector<int> &num) {
        // write you code here
        sort(num.begin(), num.end());
        map<int, int> num_map;
        int longest = 0;
        for (size_t i = 0; i < num.size(); ++ i) {
            num_map[num[i]] = num_map[num[i] - 1] + 1;
            longest = max(longest, num_map[num[i]]);
        }
        return longest;
    }
};

File: answers/leetcode/Longest Consecutive Sequence/Longest Consecutive Sequence.java
class Solution {
    public int longestConsecutive(int[] nums) {
        Map<Integer, Integer> numMap = new HashMap<>();
        for (int num : nums) {
            numMap.put(num, -1);
        }
        int maxLength = 0;
        for (int num : numMap.keySet()) {
            if (numMap.get(num) == -1) {
                this.dfs(num, numMap);
                maxLength = Math.max(maxLength, numMap.get(num));
            }
        }
        return maxLength;
    }
    
    private int dfs(int num, Map<Integer, Integer> numMap) {
        if (numMap.containsKey(num)) {
            if (numMap.get(num) == -1) {
                int next = this.dfs(num + 1, numMap);
                numMap.put(num, next + 1);
            }
            return numMap.get(num);
        }
        return 0;
    }
}